benchmark/soci_utils.go

```diff
@@ -0,0 +1,163 @@
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"syscall"
+	"time"
+
+	"github.com/awslabs/soci-snapshotter/benchmark/framework"
+	"github.com/awslabs/soci-snapshotter/fs/source"
+	"github.com/containerd/containerd"
+)
+
+var (
+	outputFilePerm  fs.FileMode = 0777
+	sociContainerId             = "TEST_SOCI_CONTAINER"
+)
+
+type SociContainerdProcess struct {
+	*framework.ContainerdProcess
+}
+
+type SociProcess struct {
+	command *exec.Cmd
+	address string
+	root    string
+	stdout  *os.File
+	stderr  *os.File
+}
+
+func StartSoci(
+	sociBinary string,
+	sociAddress string,
+	sociRoot string,
+	containerdAddress string,
+	configFile string,
+	outputDir string) (*SociProcess, error) {
+	sociCmd := exec.Command(sociBinary,
+		"-address", sociAddress,
+		"-config", configFile,
+		"-log-level", "debug",
+		"-root", sociRoot)
+	err := os.MkdirAll(outputDir, outputFilePerm)
+	if err != nil {
+		return nil, err
+	}
+	stdoutFile, err := os.Create(outputDir + "/soci-snapshotter-stdout")
+	if err != nil {
+		return nil, err
+	}
+	sociCmd.Stdout = stdoutFile
+	stderrFile, err := os.Create(outputDir + "/soci-snapshotter-stderr")
+	if err != nil {
+		return nil, err
+	}
+	sociCmd.Stderr = stderrFile
+	err = sociCmd.Start()
+	if err != nil {
+		fmt.Printf("Soci Failed to Start %v\n", err)
+		return nil, err
+	}
+
+	// The soci-snapshotter-grpc is not ready to be used until the
+	// unix socket file is created
+	sleepCount := 0
+	loopExit := false
+	for loopExit == false {
+		time.Sleep(1 * time.Second)
+		sleepCount += 1
+		if _, err := os.Stat(sociAddress); err == nil {
+			loopExit = true
+		}
+		if sleepCount > 15 {
+			return nil, errors.New("Could not create .sock in time")
+		}
+	}
+
+	return &SociProcess{
+		command: sociCmd,
+		address: sociAddress,
+		root:    sociRoot,
+		stdout:  stdoutFile,
+		stderr:  stderrFile}, nil
+}
+
+func (proc *SociProcess) StopProcess() {
+	if proc.stdout != nil {
+		proc.stdout.Close()
+	}
+	if proc.stderr != nil {
+		proc.stderr.Close()
+	}
+	if proc.command != nil {
+		proc.command.Process.Kill()
+	}
+	err := os.RemoveAll(proc.address)
+	if err != nil {
+		fmt.Printf("Error removing Address: %v\n", err)
+	}
+
+	snapshotDir := proc.root + "/snapshotter/snapshots/"
+	snapshots, err := ioutil.ReadDir(snapshotDir)
+	if err != nil {
+		fmt.Printf("Could not read dir: %s\n", snapshotDir)
+	}
+
+	for _, s := range snapshots {
+		mountpoint := snapshotDir + s.Name() + "/fs"
+		_ = syscall.Unmount(mountpoint, syscall.MNT_FORCE)
+	}
```

Comment:
Can't we remove snapshots via containerd's API instead of directly unmounting them here? The framework shouldn't know SOCI snapshotter this much.

Reply:
I don't believe that currently works.  In general, soci-snapshotter needs to put some work in managing it's data.  Right now we do not have any APIs to remove anything.

Reply:
That's sad to know :(